src/core/storage.rs

```diff
@@ -206,27 +230,213 @@ impl<K: Hash + Clone + Eq + Debug> RecordsStorage<K> {
 		}
 	}
 
-	/// Gets a value with a specific key (this method copies a value stored)
 	// TODO: think if we need to check max_ttl and initiate expiry on `get` method
-	pub fn get<Q: ?Sized>(&self, key: &Q) -> Option<StorageEntry>
+	fn get<Q: ?Sized + Hash + Eq>(&self, key: &Q) -> Option<StorageEntry>
 	where
 		K: Borrow<Q>,
-		Q: Hash + Eq,
 	{
 		self.direct_records.get(key).cloned()
 	}
 
-	/// Size of the storage
-	pub fn len(&self) -> usize {
+	fn len(&self) -> usize {
 		self.direct_records.len()
 	}
 
-	/// Returns all keys in the storage
-	pub fn keys(&self) -> Vec<K> {
+	fn keys(&self) -> Vec<K> {
 		self.direct_records.keys().cloned().collect()
 	}
 }
 
+/// This trait is used to define a storage that can store items organised in prefixes.
+/// Prefixes are used to group elements by some characteristic. For example, to get
+/// elements that belong to some particular parachain.
+pub trait PrefixedRecordsStorage<K, P> {
+	/// Insert a prefixed entry to the storage
+	fn insert_prefix(&mut self, prefix: P, key: K, entry: StorageEntry) -> color_eyre::Result<()>;
+	/// Replaces a prefixed entry in the storage, both prefix and a key must exist
```

Comment:
What does it return ?

Reply:
Same as insert: `Ok()` on insertion or an error on duplicate entry.